### DISCUSSION TITLE
Snapshot Isolation으로 Read skew를 방지할 수 있을까요?

### DISCUSSION THREAD

**Discussion by yoondgu** [woowacourse-study/2023-db-db-deep#9] [Q&A]
오늘 스터디에서 추측과 함께 질문을 던졌다가 모두를 혼란스럽게 한 질문...
동시성 제어 기법, MVCC 학습 후 정리해보기 위해 Q&A 남겨둡니당

- Snapshot isolation으로 Read skew를 방지할 수 있을까?
- Snapshot isolation은 읽기 작업에도 적용되나?
- Snapshot isolation이 읽기도 버전을 제어한다면, 최신이 아닌 버전을 읽는 경우(해당 트랜잭션보다 늦게 시작한 트랜잭션의 변경 사항이 커밋된 경우) 일관성 있는 데이터라고 할 수 있나? 
    - 그런 경우, 읽을 때까지는 문제가 없지만 쓰기를 할 수 없게 제어해주지 않을까? 먼저 커밋된 게 우선이므로..

**Comment by yeonkkk** [comment]
### Snapshot isolation은 읽기 작업에도 적용되는가?
[해당 자료](https://www.geeksforgeeks.org/what-is-snapshot-isolation/)를 보니, write-write  시에만 발생하는 것 같아요. 
혹시 제가 잘못 이해한거면 답글 주세요 ㅎㅎ
> But a transaction writing data might block other transactions trying to write data at the same time, this is an update conflict that occurs in snapshot isolation. As a result of this update conflict, any one of the two transactions needs to be rollback or it needs to be killed.

> Read-write operations will not block each other but still blocking can occur among two write operations (known as update conflicts).

**Comment by yeonkkk** [comment]
### Snapshot isolation으로 Read skew를 방지할 수 있을까? 
우선 아래 그림을 통해 강의에서 예를 들어준 read skew 예를 보면, 두 통장에서 돈을 이체하고 입금하는 예이기 때문에 전체 총량은 변하지 않아야합니다. 
하지만 트랜잭션 2가 x와 y를 읽는 사이에 시간 차가 발생하며 트랜잭션 1의 쓰기 작업이 진행되어 문제가 발생합니다.  

![image](https://github.com/woowacourse-study/2023-db-db-deep/assets/88660886/973e9612-6583-4ac8-b5e7-205d7f0db555)

만약 스냅샷을 이용한다면 특정 시점의 값을 통해 작업하므로 이와 같은 문제가 발생하지 않을 것 같습니다.  
(write - read 작업은 허용되기 때문에 읽기 작업이 rollback 되지 않습니다)  

스냅샷을 사용한다면 트랜잭션 1는 x=10, y=90, 그 전에 읽기를 시작한 트랜잭션2는 x=50, y=50 으로 알고 있을 것이고, 총량에는 변화가 없겠네요!
결론적으론.. 방지가 가능하다! 인 것 같습니다.

혹시 제가 질문을 잘못 이해한거라면 알려주세요 ㅎㅎ!!!

**Reply by yoondgu** [reply]
와우 조이선생님 너무나 야무진 설명 감사합니다.. 👍👍👍  저도 이렇게 생각했어요! (진짜임ㅎ)
